File: packages/sdk/shared/types/execution-entity.ts
/**
 * IExecutionEntity - Unified interface for all execution entity types
 *
 * Defines the common contract that both AgentLoopEntity and WorkflowExecutionEntity
 * must implement. This provides compile-time guarantees for method consistency
 * across execution instance types, replacing the previous implicit convergence
 * via union types.
 *
 * Design Principles:
 * - Implements Abortable for unified cancellation mechanism
 * - Covers lifecycle control, interruption, hierarchy, and resource cleanup
 * - Uses discriminated union type property for type-safe dispatch
 * - Does NOT include entity-type-specific methods (e.g., steering, variable management)
 *
 * @see AgentLoopEntity - Agent loop execution entity
 * @see WorkflowExecutionEntity - Workflow execution entity
 * @see ExecutionInstance - Union type using this interface for type narrowing
 */

import type {
  ParentExecutionContext,
  ChildExecutionReference,
  ExecutionHierarchyMetadata,
  ID,
} from "@wf-agent/types";
import type { ExecutionInstanceType } from "./execution.js";

/**
 * Interface for components that can be aborted/cancelled
 *
 * Provides a unified cancellation mechanism for execution entities.
 * Separated from StateManager as cancellation is an orthogonal concern:
 * state managers manage data, while abortable entities manage execution flow.
 */
export interface Abortable {
  /**
   * Abort execution with an optional reason
   * @param reason Optional reason for abortion
   */
  abort(reason?: string): void;

  /**
   * Get the AbortSignal associated with this component
   * Can be used to propagate cancellation to child operations
   */
  getAbortSignal(): AbortSignal;

  /**
   * Whether this component has been aborted
   */
  readonly aborted: boolean;
}

/**
 * Unified execution status type
 *
 * Combines statuses from both AgentLoopEntity and WorkflowExecutionEntity.
 * This is a superset of all possible statuses across both domains.
 *
 * Semantic Differences:
 * - Agent uses CANCELLED for both user-cancel and forced-stop scenarios,
 *   while Workflow distinguishes between CANCELLED (user-initiated) and
 *   STOPPED (system-initiated, e.g., timeout or error).
 * - STOPPED is only used in the Workflow domain; Agent domain does not
 *   emit this status.
 * - TIMEOUT is reserved for future use across both domains.
 *
 * When adding cross-domain cascade logic (e.g., a workflow cancelling a
 * child agent loop), the caller must map between these statuses explicitly.
 */
export type ExecutionStatus =
  | "CREATED"
  | "RUNNING"
  | "PAUSED"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED"
  | "STOPPED"
  | "TIMEOUT";

/**
 * IExecutionEntity - Common interface for all execution entities
 *
 * This interface captures the shared contract between AgentLoopEntity
 * and WorkflowExecutionEntity, providing compile-time enforcement
 * of method consistency.
 */
export interface IExecutionEntity extends Abortable {
  // ============================================================
  // Identity & Type
  // ============================================================

  /** Unique execution instance ID */
  readonly id: string;

  /** Discriminant property for type-safe dispatch */
  readonly instanceType: ExecutionInstanceType;

  // ============================================================
  // Status Access
  // ============================================================

  /** Get current execution status */
  getStatus(): ExecutionStatus;

  /** Check if currently running */
  isRunning(): boolean;

  /** Check if currently paused */
  isPaused(): boolean;

  /** Check if completed */
  isCompleted(): boolean;

  /** Check if failed */
  isFailed(): boolean;

  /** Check if cancelled */
  isCancelled(): boolean;

  // ============================================================
  // Lifecycle Control
  // ============================================================

  /** Pause execution */
  pause(): void;

  /** Resume execution */
  resume(): void;

  /** Stop/cancel execution */
  stop(): void;

  /**
   * Interrupt execution
   * @param type Interrupt type (PAUSE or STOP)
   *
   * When InterruptionState is available:
   * - Delegates to InterruptionState for cascade propagation
   * - Falls back to direct abort if InterruptionState is not set
   */
  interrupt(type: "PAUSE" | "STOP"): void;

  /** Reset interrupt flags and propagate resume to children */
  resetInterrupt(): void;

  /**
   * Check if should pause
   * Checks both internal state and InterruptionState
   */
  shouldPause(): boolean;

  /**
   * Check if should stop
   * Checks both internal state and InterruptionState
   */
  shouldStop(): boolean;

  // ============================================================
  // Hierarchy Management (Unified API)
  // ============================================================

  /** Set parent execution context */
  setParentContext(parentContext: ParentExecutionContext): void;

  /** Get parent execution context */
  getParentContext(): ParentExecutionContext | undefined;

  /** Register child execution reference */
  registerChild(childRef: ChildExecutionReference): void;

  /**
   * Unregister child execution reference
   * @param childId Child execution ID
   * @param childType Child execution type
   */
  unregisterChild(childId: ID, childType: "WORKFLOW" | "AGENT_LOOP"): void;

  /** Get all child execution references */
  getChildReferences(): ChildExecutionReference[];

  /** Get hierarchy depth (0 for root) */
  getHierarchyDepth(): number;

  /** Get root execution ID */
  getRootExecutionId(): ID;

  /** Get root execution type */
  getRootExecutionType(): "WORKFLOW" | "AGENT_LOOP";

  /** Check if this is a root execution (no parent) */
  isRootExecution(): boolean;

  /** Get hierarchy metadata for serialization */
  getHierarchyMetadata(): ExecutionHierarchyMetadata | undefined;

  /**
   * Restore hierarchy from metadata (for checkpoint restoration)
   * @param metadata Hierarchy metadata
   */
  restoreHierarchy(metadata: ExecutionHierarchyMetadata): void;

  // ============================================================
  // Resource Management
  // ============================================================

  /** Cleanup all resources held by this entity */
  cleanup(): void;
}

File: crates/engine/wf-execution-shared/tests/execution_entity.rs
use std::sync::Arc;

use async_trait::async_trait;
use wf_execution_shared::error::ExecutionSharedError;
use wf_execution_shared::types::execution_entity::{
    child_ancestors, child_depth, child_root, ExecutionEntity, ExecutionStatus,
};

struct StubEntity {
    id: String,
    status: ExecutionStatus,
    depth: u32,
    ancestors: Vec<String>,
    root: Option<String>,
}

fn stub(
    id: &str,
    status: ExecutionStatus,
    depth: u32,
    ancestors: Vec<&str>,
    root: Option<&str>,
) -> StubEntity {
    StubEntity {
        id: id.to_string(),
        status,
        depth,
        ancestors: ancestors.into_iter().map(str::to_string).collect(),
        root: root.map(str::to_string),
    }
}

#[async_trait]
impl ExecutionEntity for StubEntity {
    fn id(&self) -> &String {
        &self.id
    }

    fn status(&self) -> ExecutionStatus {
        self.status.clone()
    }

    fn is_running(&self) -> bool {
        self.status == ExecutionStatus::Running
    }

    fn is_paused(&self) -> bool {
        self.status == ExecutionStatus::Paused
    }

    fn is_completed(&self) -> bool {
        self.status == ExecutionStatus::Completed
    }

    fn is_failed(&self) -> bool {
        self.status == ExecutionStatus::Failed
    }

    fn is_cancelled(&self) -> bool {
        self.status == ExecutionStatus::Cancelled
    }

    async fn pause(&self) -> Result<(), ExecutionSharedError> {
        Ok(())
    }

    async fn resume(&self) -> Result<(), ExecutionSharedError> {
        Ok(())
    }

    async fn stop(&self) -> Result<(), ExecutionSharedError> {
        Ok(())
    }

    async fn abort(&self) {}

    fn get_abort_signal(&self) -> tokio_util::sync::CancellationToken {
        tokio_util::sync::CancellationToken::new()
    }

    fn get_hierarchy_depth(&self) -> u32 {
        self.depth
    }

    fn get_root_execution_id(&self) -> Option<String> {
        self.root.clone()
    }

    fn get_ancestors(&self) -> Vec<String> {
        self.ancestors.clone()
    }
}

#[test]
fn status_wire_names_match_serde_shape() {
    assert_eq!(ExecutionStatus::Created.as_str(), "Created");
    assert_eq!(ExecutionStatus::Running.as_str(), "Running");
    assert_eq!(ExecutionStatus::Paused.as_str(), "Paused");
    assert_eq!(ExecutionStatus::Completed.as_str(), "Completed");
    assert_eq!(ExecutionStatus::Failed.as_str(), "Failed");
    assert_eq!(ExecutionStatus::Cancelled.as_str(), "Cancelled");
    assert_eq!(ExecutionStatus::Stopped.as_str(), "Stopped");
    assert_eq!(ExecutionStatus::Timeout.as_str(), "Timeout");
}

#[test]
fn status_terminal_classification() {
    for status in [
        ExecutionStatus::Completed,
        ExecutionStatus::Failed,
        ExecutionStatus::Cancelled,
        ExecutionStatus::Stopped,
        ExecutionStatus::Timeout,
    ] {
        assert!(status.is_terminal(), "{status:?} is terminal");
    }
    for status in [
        ExecutionStatus::Created,
        ExecutionStatus::Running,
        ExecutionStatus::Paused,
    ] {
        assert!(!status.is_terminal(), "{status:?} is not terminal");
    }
}

#[test]
fn status_converts_to_persisted_shape_with_timeout_collapse() {
    assert_eq!(
        wf_types::ExecutionStatus::from(ExecutionStatus::Timeout),
        wf_types::ExecutionStatus::Failed
    );
    assert_eq!(
        wf_types::ExecutionStatus::from(ExecutionStatus::Completed),
        wf_types::ExecutionStatus::Completed
    );
    assert_eq!(
        wf_types::ExecutionStatus::from(ExecutionStatus::Running),
        wf_types::ExecutionStatus::Running
    );
}

#[test]
fn child_helpers_extend_parent_chain() {
    let parent = stub("p", ExecutionStatus::Running, 1, vec!["root"], Some("root"));
    assert_eq!(
        child_ancestors(&parent),
        vec!["root".to_string(), "p".to_string()]
    );
    assert_eq!(child_depth(&parent), 2);
    assert_eq!(child_root(&parent), "root".to_string());
}

#[test]
fn child_helpers_dedupe_tail_and_fall_back_to_parent_id() {
    let parent = stub("p", ExecutionStatus::Running, 0, vec!["p"], Some("root"));
    assert_eq!(child_ancestors(&parent), vec!["p".to_string()]);

    let orphan = stub("p", ExecutionStatus::Running, 0, vec![], None);
    assert_eq!(child_root(&orphan), "p".to_string());
}

#[tokio::test]
async fn arc_forwarding_delegates_control_plane() {
    let entity = Arc::new(stub("p", ExecutionStatus::Running, 0, vec![], None));
    assert_eq!(entity.id(), "p");
    assert!(entity.is_running());
    assert_eq!(entity.get_hierarchy_depth(), 0);
    entity.pause().await.expect("pause forwards");
    entity.resume().await.expect("resume forwards");
    entity.stop().await.expect("stop forwards");
    entity.abort().await;
}
